### PR TITLE
adds game datums

### PR DESCRIPTION
**Pull request by MadmanMartian**
To be specific, arcade datums

removes the hardcoded game from the arcade machines, ports it to arcade datums instead.

Will hopefully move to making some new games for the arcades should this succeed.

### CODE REVIEW

File: vgstation13.dme [hunk 1]
```diff
@@ -2298,7 +2299,7 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_tiny.dm"
```
Comment (Mothblocks) [inline vgstation13.dme]: please keep this i hate changing it to test_tiny

Comment (jknpj) [reply vgstation13.dme]: every time

File: code/game/machinery/computer/arcade/arcade.dm [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	circuit = "/obj/item/weapon/circuitboard/arcade"
+	var/datum/arcade_game/game
+	machine_flags = EMAGGABLE | SCREWTOGGLE | CROWDESTROY | WRENCHMOVE | FIXED2WORK
+	emag_cost = 0 // because fun
+	light_color = LIGHT_COLOR_GREEN
+
+
+/obj/machinery/computer/arcade/New()
+	..()
+	game = new /datum/arcade_game/space_villain(src)
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade.dm:15]: You need to qdel/null this in `Destroy()`

File: code/game/machinery/computer/arcade/arcade.dm [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+/obj/machinery/computer/arcade/proc/export_game_data(var/obj/item/weapon/circuitboard/arcade/A)
+	if(!A)
+		return
+	if(!A.game_data)
+		A.game_data = list()
+	A.game_data.len = 0
+	A.game_data = game.export_data()
+
+/obj/machinery/computer/arcade/attack_ai(mob/user)
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade.dm]: The parent already does this.

File: code/game/machinery/computer/arcade/arcade.dm [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	src.add_hiddenprint(user)
+	return src.attack_hand(user)
+
+/obj/machinery/computer/arcade/attack_paw(mob/user)
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade.dm]: The parent already does this.

File: code/game/machinery/computer/arcade/arcade_game.dm [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+
+/datum/arcade_game/proc/emp_act(var/severity)
+
+/datum/arcade_game/proc/kick_act()
+
+/datum/arcade_game/proc/npc_tamper_act(mob/living/L)
+
+/datum/arcade_game/New(var/holder)
+	..()
+	src.holder = holder
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade_game.dm:42]: You need to null this in `Destroy()`

File: code/game/machinery/computer/arcade/arcade_game.dm [hunk 6]
```diff
@@ -0,0 +1,359 @@
+/datum/arcade_game
+	var/name = "arcade game"
+	var/obj/machinery/computer/arcade/holder
+	var/list/prizes = list()
+	var/list/cheaters = list()
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade_game.dm:5]: You need to null this in `Destroy()`

File: code/game/machinery/computer/arcade/arcade_game.dm [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+		if(holder.stat & (NOPOWER|BROKEN))
+			return 0
+		else if(user in cheaters)
+			to_chat(usr, "<span class='danger'>[enemy_name] throws a bomb at you for trying to cheat him again.</span>")
+			explosion(get_turf(holder.loc),-1,0,2)//IED sized explosion
+			user.gib()
+			cheaters = null
+			qdel(src)
+			return 1
+		else
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade_game.dm:305]: I don't expect a proc called is_cheater to add the user to the list of cheaters but I'm also not sure this is worth changing

File: code/game/machinery/computer/arcade/arcade_game.dm [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+				log_game("[key_name_admin(usr)] has outbombed Cuban Pete and been awarded explosive maracas.")
+				holder.New()
+				emagged = 0
+
+			else if(!holder.contents.len)
+				feedback_inc("arcade_win_normal")
+				var/prizeselect = pickweight(prizes)
+				new prizeselect(holder.loc)
+
+				if(istype(prizeselect, /obj/item/toy/gun)) //Ammo comes with the gun
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade_game.dm]: Alternatively you could make prizes a list of lists and spawn both the gun and the ammo like that.
Ditto below.

File: code/game/machinery/computer/arcade/arcade_game.dm [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+	enemy_hp -= attackamt
+	arcade_action()
+
+/datum/arcade_game/space_villain/is_cheater(mob/user)
+	if(emagged && !gameover)
+		if(holder.stat & (NOPOWER|BROKEN))
+			return 0
+		else if(user in cheaters)
+			to_chat(usr, "<span class='danger'>[enemy_name] throws a bomb at you for trying to cheat him again.</span>")
+			explosion(get_turf(holder.loc),-1,0,2)//IED sized explosion
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade_game.dm]: Change `get_turf(holder.loc)` -> `holder.loc`

**Review by DamianX — CHANGES_REQUESTED** [review]

File: code/game/machinery/computer/arcade/arcade_game.dm (hunk 8 above)
Comment (MadmanMartian) [reply code/game/machinery/computer/arcade/arcade_game.dm]: done. Turns out the way they've handled that previously was having the syndicate fakespacesuit and fakespacehelmet spawn in a box.

File: code/game/machinery/computer/arcade/arcade.dm (hunk 2 above)
Comment (MadmanMartian) [reply code/game/machinery/computer/arcade/arcade.dm:15]: done

File: code/game/machinery/computer/arcade/arcade_game.dm (hunk 5 above)
Comment (MadmanMartian) [reply code/game/machinery/computer/arcade/arcade_game.dm:42]: done

File: code/game/machinery/computer/arcade/arcade_game.dm (hunk 6 above)
Comment (MadmanMartian) [reply code/game/machinery/computer/arcade/arcade_game.dm:5]: done

File: code/game/machinery/computer/arcade/arcade_game.dm (hunk 7 above)
Comment (MadmanMartian) [reply code/game/machinery/computer/arcade/arcade_game.dm:305]: It did this in the previous space villain itteration. It's an odd thing, and I'll probably change it when I add the second game.

File: code/game/machinery/computer/arcade/arcade_game.dm [hunk 10]
```diff
@@ -5,6 +5,11 @@
 	var/list/cheaters = list()
 	var/emagged = 0
 
+/datum/arcade_game/Destroy()
+	cheaters = list()
```
Comment (DamianX) [inline code/game/machinery/computer/arcade/arcade_game.dm]: Why are you creating a new list instead of nulling it?

**Comment by DamianX** [comment]
You know, some people might have liked the box for the two rewards. It was the red/syndie kind, iirc.

**Comment by MadmanMartian** [comment]
???
The box still spawns. The syndie suit wasn't in the rewards, the box that has the syndie suit is.

**Comment by DamianX** [comment]
Oh...

**Comment by MadmanMartian** [comment]
@DamianX anything still need fixing?

**Review by DamianX — APPROVED** [review]

**Pull request merged**